replica/replica.go

```diff
@@ -149,34 +157,110 @@ func (replica *Replica) Start() {
 		replica.options.Logger.Warnf("bad message: expected shard=%v, got shard=%v", replica.shard, m.Shard)
 		return
 	}
 
+	// Ignore messagse from heights that the process has already progressed
+	// through. Messages at these earlier heights have no affect on consensus,
+	// and so there is no point wasting time processing them.
+	if m.Message.Height() < replica.p.CurrentHeight() {
+		if _, ok := m.Message.(*process.Resync); !ok {
+			replica.options.Logger.Debugf("ignore message: expected height>=%v, got height=%v", replica.p.CurrentHeight(), m.Message.Height())
+			return
+		}
+	}
+	if m.Message.Type() == process.ResyncMessageType {
+		if m.Message.Height() > replica.p.CurrentHeight() {
+			// We cannot respond to resync messages from future heights with
+			// anything that is useful, so we ignore it.
+			replica.options.Logger.Debugf("ignore message: resync height=%v compared to current height=%v", m.Message.Height(), replica.p.CurrentHeight())
+			return
+		}
+		// Filter resync messages by timestamp. If they're too old, or too far
+		// in the future, then ignore them. The total window of time is 20
+		// seconds, approximately the latency expected for globally distributed
+		// message passing.
+		now := block.Timestamp(time.Now().Unix())
+		timestamp := m.Message.(*process.Resync).Timestamp()
+		delta := now - timestamp
+		if delta < 0 {
+			delta = -delta
+		}
+		if delta > 10 {
+			replica.options.Logger.Debugf("ignore message: resync timestamp=%v compared to now=%v", timestamp, now)
+			return
+		}
+	}
+
 	// Check that the Message sender is from our Shard (this can be a moderately
 	// expensive operation, so we cache the result until a new `block.Base` is
 	// detected)
 	replica.cache.fillBaseBlock(replica.blockStorage.LatestBaseBlock(replica.shard))
 	if !replica.cache.signatoryInBaseBlock(m.Message.Signatory()) {
 		return
 	}
-
 	// Verify that the Message is actually signed by the claimed `id.Signatory`
 	if err := process.Verify(m.Message); err != nil {
 		replica.options.Logger.Warnf("bad message: unverified: %v", err)
 		return
 	}
 
-	// Handle the underlying `process.Message` and immediately save the
-	// `process.Process` afterwards to protect against unexpected crashes
-	replica.p.HandleMessage(m.Message)
-	replica.p.Save()
+	// Make sure that the Process state gets saved.
+	defer replica.p.Save()
+
+	// Messages from the current height can be handled immediately.
+	if m.Message.Height() == replica.p.CurrentHeight() {
+		replica.p.HandleMessage(m.Message)
+	}
+
+	// Messages from the future must be put into the height-ordered message
+	// queue.
+	if m.Message.Height() > replica.p.CurrentHeight() {
+		if m.Message.Type() != process.ProposeMessageType {
+			// We only want to queue non-Propose messages, because Propose messages
+			// can have a LatestCommit message to fast-forward the process.
+			replica.messageQueue.Push(m.Message)
+		} else {
+			// If the Propose is not at the next height, then we need to make
```

Comment:
Should this say "If the Propose is at a future height"? Currently doesn't seem to align with the conditions in the `if` statements.

Reply:
Fixed, thank you.